### PR TITLE
express-notes-api | Angel Leung

### PR DESCRIPTION
**Pull request by angelsleung**
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  }
+  notes[nextId] = req.body;
+  notes[nextId].id = nextId;
+  fs.writeFile('data.json', JSON.stringify(notes, null, 2), err => {
+    if (err) {
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+      throw err;
+    }
+    res.status(201).json(notes[nextId]);
+    nextId++;
```
Comment (thebearingedge) [inline express-notes-api/index.js]: A couple things here:
- Avoid re-throwing in a server program unless there is some `catch` mechanism in place. In this case, I recommend just `console.error(err)` for debugging purposes and no `throw`.
- You will need an `else` to guarantee that the server cannot respond more than once to a given request.
- Increment your `data.nextId++` while it's attached to the data model before writing your JSON file.

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    res.status(404).json({ error: 'Cannot find note with id ' + id });
+    return;
+  }
+  fs.writeFile('data.json', JSON.stringify(notes, null, 2), err => {
+    if (err) {
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+      throw err;
+    }
+    delete notes[id];
+    res.status(204).json();
```
Comment (thebearingedge) [inline express-notes-api/index.js]: A couple things here:
- Avoid re-throwing in a server program unless there is some `catch` mechanism in place. In this case, I recommend just `console.error(err)` for debugging purposes and no `throw`.
- You will need an `else` to guarantee that the server cannot respond more than once to a given request.

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    return;
+  }
+  fs.writeFile('data.json', JSON.stringify(notes, null, 2), err => {
+    if (err) {
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+      throw err;
+    }
+    notes[id] = req.body;
+    notes[id].id = id;
+    res.status(200).json(notes[id]);
```
Comment (thebearingedge) [inline express-notes-api/index.js]: A couple things here:
- Avoid re-throwing in a server program unless there is some `catch` mechanism in place. In this case, I recommend just `console.error(err)` for debugging purposes and no `throw`.
- You will need an `else` to guarantee that the server cannot respond more than once to a given request.

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: express-notes-api/index.js (hunk 1 above)
Comment (angelsleung) [reply express-notes-api/index.js]: Okay, that makes sense, thank you. I'll push my changes in just a bit!

**Comment by angelsleung** [comment]
image
image

**Review by thebearingedge — APPROVED** [review]

**Pull request merged**